pkg/runner/test_runner.go

```diff
@@ -6,282 +6,796 @@
+	tr.log.Debugf("== test runner finished job %d, run %d, err: %v", jobID, runID, err)
+	return resumeState, err
 }
 
-func (w *targetWriter) writeTimeout(ctx context.Context, ch chan<- *target.Target, target *target.Target, timeout time.Duration) error {
-	w.log.Debugf("writing target %+v, timeout %v", target, timeout)
-	start := time.Now()
+func (tr *TestRunner) run(
+	ctx statectx.Context,
+	t *test.Test, targets []*target.Target,
+	jobID types.JobID, runID types.RunID,
+	resumeState []byte) ([]byte, error) {
+
+	// Peel off contexts used for steps and target handlers.
+	stepCtx, _, stepCancel := statectx.WithParent(ctx)
+	defer stepCancel()
+	targetCtx, _, targetCancel := statectx.WithParent(ctx)
+	defer targetCancel()
+
+	// Set up the pipeline
+	for i, sb := range t.TestStepsBundles {
+		tr.steps = append(tr.steps, &stepState{
+			stepIndex: i,
+			sb:        sb,
+			inCh:      make(chan *target.Target),
+			outCh:     make(chan *target.Target),
+			errCh:     make(chan cerrors.TargetError),
+			ev: storage.NewTestEventEmitter(testevent.Header{
+				JobID:         jobID,
+				RunID:         runID,
+				TestName:      t.Name,
+				TestStepLabel: sb.TestStepLabel,
+			}),
+			tgtDone: make(map[*target.Target]bool),
+			log:     logging.AddField(tr.log, "step", sb.TestStepLabel),
+		})
+		// Step handlers will be started from target handlers as targets reach them.
+	}
+
+	// Set up the targets
+	tr.targets = make(map[string]*targetState)
+	// If we have target state to resume, do it now.
+	if len(resumeState) > 0 {
+		tr.log.Debugf("Attempting to resume from state: %s", string(resumeState))
+		var rs resumeStateStruct
+		if err := json.Unmarshal(resumeState, &rs); err != nil {
+			return nil, fmt.Errorf("invalid resume state: %w", err)
+		}
+		if rs.Version != resumeStateStructVersion {
+			return nil, fmt.Errorf("incompatible resume state version %d (want %d)",
+				rs.Version, resumeStateStructVersion)
+		}
+		if rs.JobID != jobID {
+			return nil, fmt.Errorf("wrong resume state, job id %d (want %d)", rs.JobID, jobID)
+		}
+		tr.targets = rs.Targets
+	}
+	// Initialize remaining fields of the target structures,
+	// build the map and kick off target processing.
+	for _, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		if ts == nil {
+			ts = &targetState{
+				CurPhase: targetStepPhaseInit,
+			}
+		}
+		ts.tgt = tgt
+		ts.resCh = make(chan error)
+		tr.mu.Lock()
+		tr.targets[tgt.ID] = ts
+		tr.mu.Unlock()
+		tr.targetsWg.Add(1)
+		go func() {
+			tr.targetHandler(targetCtx, stepCtx, ts)
+			tr.targetsWg.Done()
+		}()
+	}
+
+	// Run until no more progress can be made.
+	runErr := tr.runMonitor()
+	if runErr != nil {
+		tr.log.Errorf("monitor returned error: %q, canceling", runErr)
+		stepCancel()
+	}
+
+	// Wait for step runners and readers to exit.
+	if err := tr.waitStepRunners(ctx); err != nil {
+		tr.log.Errorf("step runner error: %q, canceling", err)
+		stepCancel()
```

Comment:
Does `waitStepRunners` guarantee that all steps have returned? If so, `stepCancel` should not be necessary. If not, steps should be awaited for if a cancellation signal is being propagated. We should keep track of any step that does not return within the timeout and flag it accordingly.

Reply:
you are right, this is not needed as by this point all the step runners have either exited or we gave up waiting on them to exit.